lib/git_service/branch.rb

```diff
@@ -53,8 +53,24 @@ def target_for_reference(reference) # Rugged::Commit for a given refname i.e. "r
       rugged_repo.references[reference].target
     end
 
+    def tip_files
+      recursive_list_files_in_tree(tip_tree.oid)
+    end
+
     private
 
+    def recursive_list_files_in_tree(rugged_tree_oid, files = [], current_path = Pathname.new(""))
```

Comment:
I think there's way to write this where the files doesn't have to be passed all the way into the recursion.  That is, make the method more functional-like in that it returns the Array for a given path, and then it's just concating Arrays up the stack...something like:

```ruby
def recursive_list_files_in_tree(rugged_tree_oid, current_path = Pathname.new(""))
  rugged_repo.lookup(rugged_tree_oid).each_with_object([]) do |i, files|
    next_path = current_path.join(i[:name])
    case i[:type]
    when :blob
      files << next_path.to_s
    when :tree
      files.concat(recursive_list_files_in_tree(i[:oid], next_path))
    end
  end
end
```

Reply:
Upon re-reading what I wrote, you can probably remove the `recursive_` bit, since that's an implementation detail.